File: gateway/service_discovery.py
import requests
from consul import Consul

consul_url = "http://localhost:8500/v1/agent/service/register"

gateway = {
    "ID": "gateway",
    "Name": "gateway",
    "Address": "localhost",
    "Port": 5000,
}

svc1 = {
    "ID": "user-grpc-service",
    "Name": "user-grpc-service",
    "Address": "localhost",
    "Port": 50051,
}

svc2 = {
    "ID": "journal-grpc-service",
    "Name": "journal-grpc-service",
    "Address": "localhost",
    "Port": 50052,
}

def register_services():
    requests.put(consul_url, json=gateway)
    requests.put(consul_url, json=svc1)
    requests.put(consul_url, json=svc2)

consul = Consul()

def discover_grpc_server(service_name):
    services = consul.agent.services()
    grpc_service = services.get(service_name)
    if grpc_service:
        grpc_address = f"{grpc_service['Address']}:{grpc_service['Port']}"
        return grpc_address
    return None

def check(svc):
    response = requests.get(f"http://localhost:8500/v1/health/service/{svc}")
    return response.json()

File: journal_svc/internal/utils/service_discovery.go
package utils

import (
	"strconv"

	"github.com/catness812/PAD-lab1/journal_svc/internal/config"
	"github.com/catness812/PAD-lab1/journal_svc/internal/pb"
	"github.com/gookit/slog"
	"github.com/hashicorp/consul/api"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

func createConsulClient() (*api.Client, error) {
	config := api.DefaultConfig()
	config.Address = "localhost:8500"
	return api.NewClient(config)
}

func RegisterService() {
	client, err := createConsulClient()
	if err != nil {
		slog.Error(err)
		panic(err)
	}

	registration := new(api.AgentServiceRegistration)
	registration.ID = "journal-grpc-svc"
	registration.Name = "journal-grpc-svc"
	registration.Address = config.Cfg.Host
	registration.Port = config.Cfg.GrpcPort

	err = client.Agent().ServiceRegister(registration)
	if err != nil {
		slog.Error(err)
		panic(err)
	}
}

func FindUserService(serviceName string) (pb.UserServiceClient, error) {
	client, err := createConsulClient()
	if err != nil {
		slog.Error(err)
		panic(err)
	}

	serviceEntries, _, err := client.Health().Service(serviceName, "", true, nil)
	if err != nil {
		slog.Fatalf("Error querying Consul: %v", err)
		return nil, err
	}

	if len(serviceEntries) == 0 {
		slog.Fatalf("No healthy instances of %s found in Consul", serviceName)
		return nil, err
	}

	selectedService := serviceEntries[0]

	conn, err := grpc.Dial(selectedService.Service.Address+":"+strconv.Itoa(selectedService.Service.Port), grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		slog.Fatalf("Error connecting: %v", err)
		return nil, err
	}

	svcClient := pb.NewUserServiceClient(conn)
	return svcClient, nil
}
